File: purepython/mandelbrot.py
#!/usr/bin/env python3
from pathlib import Path

i_n = 16
side = 512
scale = 0.3
center = .4 + .4j


def solve_pixel(i):
    z = 0
    x = i % side
    y = i // side
    c = (x / side * 2 * scale - scale) + (1j * (y / side * scale * 2 - scale)) + center
    for k in range(i_n):
        z = z ** 2 + c
        if abs(z) > 2:
            break
    return k / i_n


res = b''.join(bytes([int(255 * c)]) * 3 for i in range(side * side) for c in [solve_pixel(i)])
Path('out.ppm').write_bytes(f'P6 {side} {side} 255 '.encode() + res)

File: julianoarr/mandelbrot.jl
const i_n = 512
const side = 1024
const center = 0.4+0.4im
const scale = 0.3


function solvePixel(i)
    x = i % side
    y = i ÷ side
    c = (x / side * 2 * scale - scale) + (y / side * 2 * scale - scale) * 1im + center
    z = 0im
    k = 0
    while (k < i_n)
        z = z * z + c
        if (abs(z) > 2)
            break
        end
        k+=1
    end
    return 1 - k / i_n
end

open("out.ppm", "w") do io
    write(io, "P6 $side $side 255 ")
    for i in 0:side-1
        buffer = Vector{UInt8}(zeros(side * 3) )
        for j in 1:side
            idx = i * side + j
            res = UInt8(floor(solvePixel(idx) * 255))
            buffer[j * 3 - 2] = res
            buffer[j * 3 - 1] = res
            buffer[j * 3 - 0] = res
        end
        write(io, buffer)
    end
end
